### PR TITLE
Modeling Outputs

### PR DESCRIPTION
**Pull request by FrancescoSaverioZuppichini**
# What does this PR do?

This PR tries to add new model outputs to cover avoid implementing custom outputs for models that do not have attentions. Three new outputs are added:
- `BaseModelOutputWithNoAttention`
- `BaseModelOutputWithPoolingAndNoAttention`
- `ImageClassifierOutputWithNoAttention`

The docstring of the `hidden_states` and `hidden_states` shape is changed to `(batch_size, num_channels, height, width)`. The thinking behind it is models that do not output attentions are usually conv models. However, this is not quite general.

Another solution is to change the docstring of `BaseModelOutput***` to state that `attentions` will be returned only if the model has an attention mechanism.

A custom `ModelOutput` will still be needed if the returned tensors' shapes are not the ones in the docstring. One solution may be making the outputs "model aware". For example, I can define a `BaseModelOutput2D` that returns 2d tensors (`batch_size, num_channels, height, width`). Doing so will unfortunately result in a lot of new model outputs. 

Let me know your feedbacks

### CODE REVIEW

**Review by sgugger — APPROVED** [review]
Works for me! As said offline, when a model doesn't match one of these new outputs in terms of shapes of the hidden states, we should just subclass and change the docstring in the modeling file.

Make so run `make style` on your branch to make the CI green.

File: src/transformers/modeling_outputs.py [hunk 1]
```diff
@@ -78,6 +97,31 @@ class BaseModelOutputWithPooling(ModelOutput):
...
+    Base class for model's outputs that also contains a pooling of the last hidden states.
+
+    Args:
+        last_hidden_state (`torch.FloatTensor` of shape `(batch_size, num_channels, height, width)`):
+            Sequence of hidden-states at the output of the last layer of the model.
+        pooler_output (`torch.FloatTensor` of shape `(batch_size, hidden_size)`):
+            Last layer hidden-state of the first token of the sequence (classification token) after further processing
+            through the layers used for the auxiliary pretraining task. E.g. for BERT-family of models, this returns
+            the classification token after processing through a linear layer and a tanh activation function. The linear
+            layer weights are trained from the next sentence prediction (classification) objective during pretraining.
```
[the remark this answers is not in the document]
Comment (FrancescoSaverioZuppichini) [reply src/transformers/modeling_outputs.py:109]: Sure, I've updated the docstring :)

**Comment by FrancescoSaverioZuppichini** [comment]
Updated the docstring of the outputs and replace multiple models' outputs with the new ones

File: src/transformers/modeling_outputs.py [hunk 2]
```diff
@@ -46,6 +46,25 @@ class BaseModelOutput(ModelOutput):
     attentions: Optional[Tuple[torch.FloatTensor]] = None
 
 
+@dataclass
+class BaseModelOutputWithNoAttention(ModelOutput):
+    """
+    Base class for model's outputs, with potential hidden states and attentions.
```
Comment (NielsRogge) [inline src/transformers/modeling_outputs.py:52]:
```suggestion
    Base class for model's outputs, with potential hidden states.
```
This class doesn't have attention :p

File: src/transformers/modeling_outputs.py [hunk 3]
```diff
@@ -55,10 +74,7 @@ class BaseModelOutputWithPooling(ModelOutput):
         last_hidden_state (`torch.FloatTensor` of shape `(batch_size, sequence_length, hidden_size)`):
             Sequence of hidden-states at the output of the last layer of the model.
         pooler_output (`torch.FloatTensor` of shape `(batch_size, hidden_size)`):
-            Last layer hidden-state of the first token of the sequence (classification token) after further processing
-            through the layers used for the auxiliary pretraining task. E.g. for BERT-family of models, this returns
-            the classification token after processing through a linear layer and a tanh activation function. The linear
-            layer weights are trained from the next sentence prediction (classification) objective during pretraining.
+            Last layer hidden-state after a pooling operation on the spatial dimensions.
```
Comment (NielsRogge) [inline src/transformers/modeling_outputs.py:77]: Hmm that might be a bit too limited.

cc'ing @patrickvonplaten here - this change would be made to make `BaseModelOutputWithPooling` also useful for vision models, for which this docstring isn't applicable.

Comment (sgugger) [reply src/transformers/modeling_outputs.py:77]: Ah good catch, I wanted that change on the `BaseModelOutputWithNoAttentionAndWithPooling`, not this class, which should be left untouched.

File: src/transformers/modeling_outputs.py [hunk 4]
```diff
@@ -78,6 +94,28 @@ class BaseModelOutputWithPooling(ModelOutput):
     attentions: Optional[Tuple[torch.FloatTensor]] = None
 
 
+@dataclass
+class BaseModelOutputWithNoAttentionAndWithPooling(ModelOutput):
```
Comment (sgugger) [inline src/transformers/modeling_outputs.py:98]:
```suggestion
class BaseModelOutputWithPoolingAndNoAttention(ModelOutput):
```
Alternate name that would not have two with.

Comment (FrancescoSaverioZuppichini) [reply src/transformers/modeling_outputs.py:98]: okay

File: src/transformers/modeling_outputs.py (hunk 2 above)
Comment (FrancescoSaverioZuppichini) [reply src/transformers/modeling_outputs.py:52]: thanks :)

File: src/transformers/modeling_outputs.py (hunk 3 above)
Comment (FrancescoSaverioZuppichini) [reply src/transformers/modeling_outputs.py:77]: Thanks for the comments! The doc is also updated in `BaseModelOutputWithNoAttentionAndWithPooling`. We can also be more general and just say ` Last layer hidden-state after a pooling operation.`

File: src/transformers/modeling_outputs.py [hunk 5]
```diff
@@ -46,6 +46,25 @@ class BaseModelOutput(ModelOutput):
...
+class BaseModelOutputWithNoAttention(ModelOutput):
+    """
+    Base class for model's outputs, with potential hidden states.
+
+    Args:
+        last_hidden_state (`torch.FloatTensor` of shape `(batch_size, num_channels, height, width)`):
+            Sequence of hidden-states at the output of the last layer of the model.
+        hidden_states (`tuple(torch.FloatTensor)`, *optional*, returned when `output_hidden_states=True` is passed or when `config.output_hidden_states=True`):
+            Tuple of `torch.FloatTensor` (one for the output of the embeddings + one for the output of each layer) of
+            shape `(batch_size, num_channels, height, width)`.
```
Comment (NielsRogge) [inline src/transformers/modeling_outputs.py:59]: Some models, like SegFormer, don't have an initial embedding layer.

File: src/transformers/modeling_outputs.py (hunk 3 above)
Comment (NielsRogge) [reply src/transformers/modeling_outputs.py:77]: Yes but you can actually revert this change here :)

File: src/transformers/modeling_outputs.py [hunk 6]
```diff
@@ -78,6 +94,28 @@ class BaseModelOutputWithPooling(ModelOutput):
...
+    """
+    Base class for model's outputs that also contains a pooling of the last hidden states.
+
+    Args:
+        last_hidden_state (`torch.FloatTensor` of shape `(batch_size, num_channels, height, width)`):
+            Sequence of hidden-states at the output of the last layer of the model.
+        pooler_output (`torch.FloatTensor` of shape `(batch_size, hidden_size)`):
+            Last layer hidden-state after a pooling operation on the spatial dimensions.
+        hidden_states (`tuple(torch.FloatTensor)`, *optional*, returned when `output_hidden_states=True` is passed or when `config.output_hidden_states=True`):
+            Tuple of `torch.FloatTensor` (one for the output of the embeddings + one for the output of each layer) of
```
Comment (NielsRogge) [inline src/transformers/modeling_outputs.py:108]: Same comment here, some models like SegFormer don't have an initial embedding layer.

File: src/transformers/modeling_outputs.py [hunk 7]
```diff
@@ -880,6 +918,27 @@ class ImageClassifierOutput(ModelOutput):
...
+    """
+    Base class for outputs of image classification models.
+
+    Args:
+        loss (`torch.FloatTensor` of shape `(1,)`, *optional*, returned when `labels` is provided):
+            Classification (or regression if config.num_labels==1) loss.
+        logits (`torch.FloatTensor` of shape `(batch_size, config.num_labels)`):
+            Classification (or regression if config.num_labels==1) scores (before SoftMax).
+        hidden_states (`tuple(torch.FloatTensor)`, *optional*, returned when `output_hidden_states=True` is passed or when `config.output_hidden_states=True`):
+            Tuple of `torch.FloatTensor` (one for the output of the embeddings + one for the output of each stage) of
```
Comment (NielsRogge) [inline src/transformers/modeling_outputs.py:932]: Same comment here for models like SegFormer

**Comment by sgugger** [comment]
@NielsRogge Not sure what the point of your last comments. We know the modeling outputs won't work for every model and we have a strategy for that (subclass and overwrite the docstring). Did you have a suggestion on how to make the docstring more general?

**Comment by NielsRogge** [comment]
Maybe we can add a statement that says that the `hidden_states` optionally include the initial embeddings, if the model has an embedding layer.

File: src/transformers/modeling_outputs.py (hunk 5 above)
Comment (NielsRogge) [reply src/transformers/modeling_outputs.py:59]: Update: it's only PoolFormer

File: src/transformers/modeling_outputs.py (hunk 3 above)
Comment (FrancescoSaverioZuppichini) [reply src/transformers/modeling_outputs.py:77]: oh okay, I see! I got confused by my own names! Thanks

File: src/transformers/modeling_outputs.py (hunk 5 above)
Comment (FrancescoSaverioZuppichini) [reply src/transformers/modeling_outputs.py:59]: > Maybe we can add a statement that says that the `hidden_states` optionally include the initial embeddings, if the model has an embedding layer.

+1 for @NielsRogge idea

**Comment by FrancescoSaverioZuppichini** [comment]
+1 on @NielsRogge idea

**Comment by NielsRogge** [comment]
Looks good to me, note that we have some more vision models:

* Swin Transformer
* BEiT

File: src/transformers/modeling_outputs.py (hunk 5 above)
Comment (FrancescoSaverioZuppichini) [reply src/transformers/modeling_outputs.py:59]: VAN as well

**Comment by FrancescoSaverioZuppichini** [comment]
Added correct outputs for
- DeiT
- BEiT
The following models need a custom output
- ViTMAE
- Swin

File: src/transformers/modeling_outputs.py [hunk 8]
```diff
@@ -29,7 +29,7 @@ class BaseModelOutput(ModelOutput):
         last_hidden_state (`torch.FloatTensor` of shape `(batch_size, sequence_length, hidden_size)`):
             Sequence of hidden-states at the output of the last layer of the model.
         hidden_states (`tuple(torch.FloatTensor)`, *optional*, returned when `output_hidden_states=True` is passed or when `config.output_hidden_states=True`):
-            Tuple of `torch.FloatTensor` (one for the output of the embeddings + one for the output of each layer) of
+            Tuple of `torch.FloatTensor` (one for the output of the embeddings, if the model's has an embedding layer, + one for the output of each layer) of
```
Comment (sgugger) [inline src/transformers/modeling_outputs.py:32]:
```suggestion
            Tuple of `torch.FloatTensor` (one for the output of the embeddings, if the model has an embedding layer, + one for the output of each layer) of
```

**Comment by FrancescoSaverioZuppichini** [comment]
Changed the docstring for every `hidden_states` field in every `*ModelOutput*` to specify the embedding hidden state may be optional (depending on the model's architecture)

File: src/transformers/modeling_outputs.py [hunk 9]
```diff
@@ -29,8 +29,8 @@ class BaseModelOutput(ModelOutput):
         last_hidden_state (`torch.FloatTensor` of shape `(batch_size, sequence_length, hidden_size)`):
             Sequence of hidden-states at the output of the last layer of the model.
         hidden_states (`tuple(torch.FloatTensor)`, *optional*, returned when `output_hidden_states=True` is passed or when `config.output_hidden_states=True`):
-            Tuple of `torch.FloatTensor` (one for the output of the embeddings + one for the output of each layer) of
-            shape `(batch_size, sequence_length, hidden_size)`.
+            Tuple of `torch.FloatTensor` (one for the output of the embeddings, if the models has an embedding layer,,
```
Comment (NielsRogge) [inline src/transformers/modeling_outputs.py:32]:
```suggestion
            Tuple of `torch.FloatTensor` (one for the output of the embeddings, if the models has an embedding layer,
```

File: src/transformers/modeling_outputs.py (hunk 8 above)
Comment (FrancescoSaverioZuppichini) [reply src/transformers/modeling_outputs.py]: Thanks! Fix it

File: src/transformers/modeling_outputs.py [hunk 10]
```diff
@@ -46,6 +46,25 @@ class BaseModelOutput(ModelOutput):
...
+    Base class for model's outputs, with potential hidden states.
+
+    Args:
+        last_hidden_state (`torch.FloatTensor` of shape `(batch_size, num_channels, height, width)`):
+            Sequence of hidden-states at the output of the last layer of the model.
+        hidden_states (`tuple(torch.FloatTensor)`, *optional*, returned when `output_hidden_states=True` is passed or when `config.output_hidden_states=True`):
+            Tuple of `torch.FloatTensor` (one for the output of the embeddings, if the model has an embedding layer, +
+            one for the output of each layer) of shape `(batch_size, num_channels, height, width)`.
+
+            Hidden-states of the model at the output of each layer plus the initial embedding outputs.
```
Comment (NielsRogge) [inline src/transformers/modeling_outputs.py:61]:
```suggestion
            Hidden-states of the model at the output of each layer plus the optional initial embedding outputs.
```

File: src/transformers/modeling_outputs.py [hunk 11]
```diff
@@ -99,8 +140,8 @@ class BaseModelOutputWithPast(ModelOutput):
             `config.is_encoder_decoder=True` in the cross-attention blocks) that can be used (see `past_key_values`
             input) to speed up sequential decoding.
         hidden_states (`tuple(torch.FloatTensor)`, *optional*, returned when `output_hidden_states=True` is passed or when `config.output_hidden_states=True`):
-            Tuple of `torch.FloatTensor` (one for the output of the embeddings + one for the output of each layer) of
-            shape `(batch_size, sequence_length, hidden_size)`.
+            Tuple of `torch.FloatTensor` (one for the output of the embeddings, if the model has an embedding layer, +
+            one for the output of each layer) of shape `(batch_size, sequence_length, hidden_size)`.
 
             Hidden-states of the model at the output of each layer plus the initial embedding outputs.
```
Comment (NielsRogge) [inline src/transformers/modeling_outputs.py:146]:
```suggestion
            Hidden-states of the model at the output of each layer plus the optional initial embedding outputs.
```

File: src/transformers/modeling_outputs.py [hunk 12]
```diff
@@ -164,8 +205,8 @@ class BaseModelOutputWithPoolingAndCrossAttentions(ModelOutput):
             the classification token after processing through a linear layer and a tanh activation function. The linear
             layer weights are trained from the next sentence prediction (classification) objective during pretraining.
         hidden_states (`tuple(torch.FloatTensor)`, *optional*, returned when `output_hidden_states=True` is passed or when `config.output_hidden_states=True`):
-            Tuple of `torch.FloatTensor` (one for the output of the embeddings + one for the output of each layer) of
-            shape `(batch_size, sequence_length, hidden_size)`.
+            Tuple of `torch.FloatTensor` (one for the output of the embeddings, if the model has an embedding layer, +
+            one for the output of each layer) of shape `(batch_size, sequence_length, hidden_size)`.
 
             Hidden-states of the model at the output of each layer plus the initial embedding outputs.
```
Comment (NielsRogge) [inline src/transformers/modeling_outputs.py:211]:
```suggestion
            Hidden-states of the model at the output of each layer plus the optional initial embedding outputs.
```

File: src/transformers/modeling_outputs.py [hunk 13]
```diff
@@ -220,8 +261,8 @@ class BaseModelOutputWithPastAndCrossAttentions(ModelOutput):
             `config.is_encoder_decoder=True` in the cross-attention blocks) that can be used (see `past_key_values`
             input) to speed up sequential decoding.
         hidden_states (`tuple(torch.FloatTensor)`, *optional*, returned when `output_hidden_states=True` is passed or when `config.output_hidden_states=True`):
-            Tuple of `torch.FloatTensor` (one for the output of the embeddings + one for the output of each layer) of
-            shape `(batch_size, sequence_length, hidden_size)`.
+            Tuple of `torch.FloatTensor` (one for the output of the embeddings, if the model has an embedding layer, +
+            one for the output of each layer) of shape `(batch_size, sequence_length, hidden_size)`.
 
             Hidden-states of the model at the output of each layer plus the initial embedding outputs.
```
Comment (NielsRogge) [inline src/transformers/modeling_outputs.py:267]:
```suggestion
            Hidden-states of the model at the output of each layer plus the optional initial embedding outputs.
```

File: src/transformers/modeling_outputs.py [hunk 14]
```diff
@@ -265,8 +306,8 @@ class Seq2SeqModelOutput(ModelOutput):
             Contains pre-computed hidden-states (key and values in the self-attention blocks and in the cross-attention
             blocks) that can be used (see `past_key_values` input) to speed up sequential decoding.
         decoder_hidden_states (`tuple(torch.FloatTensor)`, *optional*, returned when `output_hidden_states=True` is passed or when `config.output_hidden_states=True`):
-            Tuple of `torch.FloatTensor` (one for the output of the embeddings + one for the output of each layer) of
-            shape `(batch_size, sequence_length, hidden_size)`.
+            Tuple of `torch.FloatTensor` (one for the output of the embeddings, if the model has an embedding layer, +
+            one for the output of each layer) of shape `(batch_size, sequence_length, hidden_size)`.
 
             Hidden-states of the decoder at the output of each layer plus the initial embedding outputs.
```
Comment (NielsRogge) [inline src/transformers/modeling_outputs.py:312]:
```suggestion
            Hidden-states of the decoder at the output of each layer plus optional the initial embedding outputs.
```

File: src/transformers/modeling_outputs.py [hunk 15]
```diff
@@ -284,8 +325,8 @@ class Seq2SeqModelOutput(ModelOutput):
         encoder_last_hidden_state (`torch.FloatTensor` of shape `(batch_size, sequence_length, hidden_size)`, *optional*):
             Sequence of hidden-states at the output of the last layer of the encoder of the model.
         encoder_hidden_states (`tuple(torch.FloatTensor)`, *optional*, returned when `output_hidden_states=True` is passed or when `config.output_hidden_states=True`):
-            Tuple of `torch.FloatTensor` (one for the output of the embeddings + one for the output of each layer) of
-            shape `(batch_size, sequence_length, hidden_size)`.
+            Tuple of `torch.FloatTensor` (one for the output of the embeddings, if the model has an embedding layer, +
+            one for the output of each layer) of shape `(batch_size, sequence_length, hidden_size)`.
 
             Hidden-states of the encoder at the output of each layer plus the initial embedding outputs.
```
Comment (NielsRogge) [inline src/transformers/modeling_outputs.py:331]:
```suggestion
            Hidden-states of the encoder at the output of each layer plus the optional initial embedding outputs.
```

File: src/transformers/modeling_outputs.py [hunk 16]
```diff
@@ -317,8 +358,8 @@ class CausalLMOutput(ModelOutput):
         logits (`torch.FloatTensor` of shape `(batch_size, sequence_length, config.vocab_size)`):
             Prediction scores of the language modeling head (scores for each vocabulary token before SoftMax).
         hidden_states (`tuple(torch.FloatTensor)`, *optional*, returned when `output_hidden_states=True` is passed or when `config.output_hidden_states=True`):
-            Tuple of `torch.FloatTensor` (one for the output of the embeddings + one for the output of each layer) of
-            shape `(batch_size, sequence_length, hidden_size)`.
+            Tuple of `torch.FloatTensor` (one for the output of the embeddings, if the model has an embedding layer, +
+            one for the output of each layer) of shape `(batch_size, sequence_length, hidden_size)`.
 
             Hidden-states of the model at the output of each layer plus the initial embedding outputs.
```
Comment (NielsRogge) [inline src/transformers/modeling_outputs.py:364]:
```suggestion
            Hidden-states of the model at the output of each layer plus the optional initial embedding outputs.
```

File: src/transformers/modeling_outputs.py [hunk 17]
```diff
@@ -352,8 +393,8 @@ class CausalLMOutputWithPast(ModelOutput):
             Contains pre-computed hidden-states (key and values in the self-attention blocks) that can be used (see
             `past_key_values` input) to speed up sequential decoding.
         hidden_states (`tuple(torch.FloatTensor)`, *optional*, returned when `output_hidden_states=True` is passed or when `config.output_hidden_states=True`):
-            Tuple of `torch.FloatTensor` (one for the output of the embeddings + one for the output of each layer) of
-            shape `(batch_size, sequence_length, hidden_size)`.
+            Tuple of `torch.FloatTensor` (one for the output of the embeddings, if the model has an embedding layer, +
+            one for the output of each layer) of shape `(batch_size, sequence_length, hidden_size)`.
 
             Hidden-states of the model at the output of each layer plus the initial embedding outputs.
```
Comment (NielsRogge) [inline src/transformers/modeling_outputs.py:399]: Same comment for all output classes below.

**Review by NielsRogge — APPROVED** [review]
LGTM! Thanks for improving this.

I think that SegFormer is the only exception for which the general docstrings won't be correct.

File: src/transformers/models/segformer/modeling_segformer.py [hunk 18]
```diff
@@ -625,7 +625,7 @@ def forward(
             output = (logits,) + outputs[1:]
             return ((loss,) + output) if loss is not None else output
 
-        return SequenceClassifierOutput(
+        return ImageClassifierOutput(
```
[the remark this answers is not in the document]
Comment (FrancescoSaverioZuppichini) [reply src/transformers/models/segformer/modeling_segformer.py]: Create `SegFormerImageClassifierOutput` :)

**Pull request merged**